### PR TITLE
Update torch::lazy::BackendDevice to have a new default ordinal

### LINKED ISSUE #3490: Use upstream `BackendDevice` instead of inherit from it
## 🚀 Feature
Currently  we inherit a Device from `torch::lazy::BackendDevice` but that should be unneceaary.
https://github.com/pytorch/xla/blob/94d71650159ea6ef958abd7f8374922c4a536705/torch_xla/csrc/device.h#L24

We should use `BackendDevice` directly in PyTorch/XLA.

## Motivation

Going forward we will need to use functions like `torch::lazy::GetBackendDevice` which returns a `c10::optional<BackendDevice>`

This is also blocking us from using code-gen.

@alanwaketan @wonjoolee95

### PR DESCRIPTION
**Pull request by wonjoo-wj**
Fixes https://github.com/pytorch/xla/issues/3490. Updates `torch::lazy::BackendDevice` with changes below:

1. Remove the no-op string constructor.
2. Update default ordinal to `-1`.
3. Add a `is_valid` function to check if `ordinal` is valid/non-default (`ordinal >= 0`).

### CODE REVIEW

File: test/cpp/lazy/test_backend_device.cpp
Comment (wonjoolee95) [inline test/cpp/lazy/test_backend_device.cpp:25]: Just FYI, the `toString` response with default ordinal (-1) now becomes `Unknown-1` (this may be a bit confusing because the negative can be interpreted as a hyphen).

File: test/cpp/lazy/test_backend_device.cpp
Comment (wonjoolee95) [inline test/cpp/lazy/test_backend_device.cpp:92]: This is due to `c10::Device` also treating `-1` has a default index/ordinal. And `c10::Device::has_index()` returns `false` if ordinal is `-1`. 

Reference: https://github.com/pytorch/pytorch/blob/master/c10/core/Device.h#L75. .

Comment (alanwaketan) [reply test/cpp/lazy/test_backend_device.cpp:92]: Can we follow the convention of c10::Device and rename is_valid to has_index?

File: test/cpp/lazy/test_backend_device.cpp
Comment (alanwaketan) [reply test/cpp/lazy/test_backend_device.cpp:25]: Can you check what torch.device get printed in python by default? We can follow the convention there.

**Review by alanwaketan — APPROVED** [review]
LGTM in general. Please wait until all CI are green before landing.

**Comment by alanwaketan** [comment]
Also, we could remove this particular line of comment:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/backend/backend_device.cpp#L13.

And try to update this method as well:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/backend/backend_device.cpp#L46.

File: test/cpp/lazy/test_backend_device.cpp
Comment (wonjoolee95) [reply test/cpp/lazy/test_backend_device.cpp]: Good point! So I've updated to follow the `c10::Device::str()` -- https://github.com/pytorch/pytorch/blob/master/c10/core/Device.cpp#L135. Ordinal only gets appended to the string if it exists.

File: test/cpp/lazy/test_backend_device.cpp
Comment (wonjoolee95) [reply test/cpp/lazy/test_backend_device.cpp:92]: Updated.

**Comment by wonjoolee95** [comment]
> Also, we could remove this particular line of comment: https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/backend/backend_device.cpp#L13.
> 
> And try to update this method as well: https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/backend/backend_device.cpp#L46.

Updated, thanks @alanwaketan!

File: torch/csrc/lazy/backend/backend_device.cpp
Comment (alanwaketan) [inline torch/csrc/lazy/backend/backend_device.cpp:45]: I guess we can also remove this comment?

**Review by alanwaketan — APPROVED** [review]

**Comment by wonjoolee95** [comment]
@alanwaketan, thanks for the comments, this should be ready to merge (last CI was all green, and the newest commit just removes a comment). However, seems like I don't have merging access.

**Comment by alanwaketan** [comment]
> @alanwaketan, thanks for the comments, this should be ready to merge. However, seems like I don't have merging access.

I will handle the rest.

**Comment by alanwaketan** [comment]
BTW, I think you can reland part of this commit just don't set the default ordinal to -1.

**Comment by JackCaoG** [comment]
@wonjoolee95

**Comment by wonjoolee95** [comment]
> BTW, I think you can reland part of this commit just don't set the default ordinal to -1.

Sounds good, let me update the PR without the -1 default ordinal.

**Comment by wonjoolee95** [comment]
Hmm, looking at this commit, I don't think there's much we can do without setting the `-1` default ordinal. The newly introduced `has_index()` function and the updated `ToString()` function rely on the default ordinal being `-1`. Seems like the only part of we can land is removing the deprecated string constructor. FYI: @alanwaketan.

**Comment by alanwaketan** [comment]
For XLA, do you need the ability to detect has_index? If so, you can construct a BackendDevice with ordinal set to -1 to test those functionalities. I'm working on a backend API for you to set the ordinal to -1.

**Comment by JackCaoG** [comment]
has_index is just to detect when deivce is init incorrectly. If we can set the `default_device_type` and `defualt_device_ordinal` we should be able to assume all device is valid.

**Comment by wonjoolee95** [comment]
Sounds good, so if I understand correctly, @alanwaketan will implement the public APIs on `torch::lazy::BackendDevice` to set the device type and ordinal, is that correct? If so, I can follow-up on XLA's end to call these APIs to make sure devices are all valid once that change is merged.

**Pull request closed without merging**
